src/Core/Domain/Carrier/QueryHandler/GetCarrierRangesHandlerInterface.php

```diff
@@ -0,0 +1,43 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShop\PrestaShop\Core\Domain\Carrier\QueryHandler;
+
+use PrestaShop\PrestaShop\Core\Domain\Carrier\Query\GetCarrierRanges;
+use PrestaShop\PrestaShop\Core\Domain\Carrier\ValueObject\CarrierRangesCollection;
+
+/**
+ * Describes get carrier ranges handler.
+ */
+interface GetCarrierRangesHandlerInterface
+{
+    /**
+     * @param GetCarrierRanges $query
+     *
+     * @return CarrierRangesCollection
+     */
+    public function handle(GetCarrierRanges $query): CarrierRangesCollection;
```

Comment:
After a bit of thinking and digging into our current commands and queries, I'm more and more sure these two structure must be separated The principle of CQRS is to separate the commands (write) from the queries (read) and we don't have the same constraint in read or write

The write command must be validated that's why they rely on VOs, the read operation don't need this since the data is supposed to be already valid and consistent So by convention we also try to keep the structure as flat as possible which eases the seralization later (especially useful for the API)

One example I could find is about the product redirection:
this is the VO used in the command https://github.com/PrestaShop/PrestaShop/blob/ada985284ce331b28539adfc3b374879e1749b2c/src/Core/Domain/Product/ValueObject/RedirectType.php
this is the Queryresult used in the query https://github.com/PrestaShop/PrestaShop/blob/131d60617c4237077ff0bf7f7fe726dfb5de63f5/src/Core/Domain/QueryResult/RedirectTargetInformation.php

Both objects actually represent the same data, but we still have two distinct objects that are used for different purposes. It also guaranties that they are not coupled and can leave and be modified indepedently.

Reply:
In a nutshell you should create some query results objects for the ranges, the difference with the VOs is that they don't need asssertions We can still create objects instead of returning a big flat array because it still help to structure our data more cleanly and with stricter types and it won't change the way it will be serialized anyway